sycl/include/sycl/ext/oneapi/experimental/group_load_store.hpp

```diff
@@ -0,0 +1,1006 @@
+                  //    6    | s1.w6.b3 | s2.w5.b0 | s2.w4.b1 | s2.w3.b2 | s2.w2.b3 |
+          // clang-format on
+          while (true) {
+            int needed_idx = needed_global_idx - cur_hw_blocks_start_idx;
+
+            int wi = needed_idx % sg_size;
+            int block = needed_idx / sg_size;
+
+            // Shuffle has to be in the convergent control flow.
+            auto val = select_from_group(sg, load, wi);
+
+            bool write_needed = needed_global_idx >= cur_hw_blocks_start_idx &&
+                                needed_global_idx < cur_hw_blocks_start_idx +
+                                                        hw_blocks_per_iter;
+
+            if (none_of_group(sg, write_needed))
+              break;
+
+            if (write_needed) {
+              std::memcpy(priv_ptr + cur_write_index * hw_block_size,
+                          reinterpret_cast<HWBlockTy *>(&val) + block,
+                          hw_block_size);
+              ++cur_write_index;
+              needed_global_idx +=
+                  cur_write_index % hw_blocks_per_block == 0
+                      ? 1 + sg_size * hw_blocks_per_block - hw_blocks_per_block
+                      : 1;
+            }
+          };
+        }
+
+        cur_hw_blocks_start_idx += hw_blocks_per_iter;
+      });
+    });
+
+    // Now perform the required implicit conversion.
+    detail::loop_unroll_up_to<ElementsPerWorkItem, 16>(
+        [&](size_t i) { out[i] = v[i]; });
+  };
+
+  auto impl = [&](auto *in_ptr) {
+    group_barrier(g);
+    if constexpr (detail::is_sub_group<GroupTy>::value) {
+      return impl_sg(g, in_ptr);
+    } else {
+      // TODO: Use get_child_group from sycl_ext_oneapi_root_group extension
+      // once it is implemented instead of this free function.
+      auto ndi =
+          sycl::ext::oneapi::experimental::this_nd_item<GroupTy::dimensions>();
+      auto sg = ndi.get_sub_group();
+      if constexpr (blocked) {
+        return impl_sg(sg, in_ptr + sg.get_group_id() *
+                                        sg.get_max_local_range() *
+                                        ElementsPerWorkItem);
+      } else {
+        // For striped layout the stride between elements in a vector is
+        // expressed in terms of WG's size, not SG. As such, each index has
+        // to be implemented using scalar SG block_load.
+        auto vec_elem_stride = g.get_local_linear_range();
+        detail::loop_unroll_up_to<ElementsPerWorkItem, 16>([&](size_t i) {
+          OutputT scalar;
+          group_load(sg,
+                     in_ptr + sg.get_group_id() * sg.get_max_local_range() +
+                         vec_elem_stride * i,
+                     scalar, properties);
+          out[i] = scalar;
+        });
+      }
+    }
+    group_barrier(g);
+  };
+
+  constexpr bool assume_full_sg =
+      properties.template has_property<property::full_sg_key>();
+  // We'd need too much private memory.
```

Comment:
I feel like this comment is trying to tell me something, but I'm not sure I understand what.